hw/drivers/chg_ctrl/da1469x_charger/src/da1469x_charger_shell.c

```diff
@@ -0,0 +1,970 @@
+    FIELD_NUM(ERROR_IRQ_CLR, CC_CHARGE_TIMEOUT_IRQ_CLR),
+    FIELD_NUM(ERROR_IRQ_CLR, PRECHARGE_TIMEOUT_IRQ_CLR),
+    { NULL }
+};
+
+#define REG(reg_name) { .addr = &CHARGER->CHARGER_##reg_name##_REG, .name = "CHARGER_"#reg_name"_REG", .fields = reg_name##_fields }
+
+#else
+#define REG(reg_name) { .addr = &CHARGER->CHARGER_##reg_name##_REG, .name = "CHARGER_"#reg_name"_REG", .fields = NULL }
+#endif
+
+#define REG_NAME(short_name) "CHARGER_"#short_name"_REG"
+
+static const struct reg charger_regs[] = {
+    REG(CTRL),
+    REG(STATUS),
+    REG(VOLTAGE_PARAM),
+    REG(CURRENT_PARAM),
+    REG(TEMPSET_PARAM),
+    REG(PRE_CHARGE_TIMER),
+    REG(CC_CHARGE_TIMER),
+    REG(CV_CHARGE_TIMER),
+    REG(TOTAL_CHARGE_TIMER),
+    REG(JEITA_V_CHARGE),
+    REG(JEITA_V_PRECHARGE),
+    REG(JEITA_V_REPLENISH),
+    REG(JEITA_V_OVP),
+    REG(JEITA_CURRENT),
+    REG(VBAT_COMP_TIMER),
+    REG(VOVP_COMP_TIMER),
+    REG(TDIE_COMP_TIMER),
+    REG(TBAT_MON_TIMER),
+    REG(TBAT_COMP_TIMER),
+    REG(THOT_COMP_TIMER),
+    REG(PWR_UP_TIMER),
+    REG(STATE_IRQ_MASK),
+    REG(ERROR_IRQ_MASK),
+    REG(STATE_IRQ_STATUS),
+    REG(ERROR_IRQ_STATUS),
+    REG(STATE_IRQ_CLR),
+    REG(ERROR_IRQ_CLR),
+};
+#define NUM_CHARGER_REGS (sizeof(charger_regs) / sizeof(charger_regs[0]))
+
+static const struct shell_cmd da1469x_charger_shell_cmd_struct;
+
+static int
+da1469x_charger_help(void)
+{
+    console_printf("%s cmd\n", da1469x_charger_shell_cmd_struct.sc_cmd);
+    console_printf("cmd:\n");
+    console_printf("\thelp\n");
+#if MYNEWT_VAL(DA1469X_CHARGER_CLI_DECODE)
+    console_printf("\tdump [decode]\n");
+#else
+    console_printf("\tdump\n");
+#endif
+    console_printf("\tread reg\n");
+    console_printf("\twrite reg\n");
+    console_printf("\tdisable\n");
+    console_printf("\tenable\n");
+    console_printf("\tstatus\n");
+    return 0;
+}
+
+static int
+da1469x_charger_shell_err_too_many_args(const char *cmd_name)
+{
+    console_printf("Error: too many arguments for command \"%s\"\n",
+                   cmd_name);
+    return SYS_EINVAL;
+}
+
+static int
+da1469x_charger_shell_err_unknown_arg(const char *cmd_name)
+{
+    console_printf("Error: unknown argument \"%s\"\n",
+                   cmd_name);
+    return SYS_EINVAL;
+}
+
+#if MYNEWT_VAL(DA1469X_CHARGER_CLI_DECODE)
+static const char *
+field_bin_value(const struct reg_field *field, uint32_t val, char *buf)
+{
+    uint32_t mask = field->fld_mask >> field->fld_pos;
+    int bits = 0;
+    val = (val & field->fld_mask) >> field->fld_pos;
+    for (bits = 0; mask; bits++) {
+        mask >>= 1;
+    }
+    mask = field->fld_mask >> field->fld_pos;
+    buf[bits] = '\0';
+    while (mask) {
+        buf[--bits] = (char)('0' + (val & 1));
+        mask >>= 1;
+        val >>= 1;
+    }
+    return buf;
+}
+
+static int
+field_int_value(const struct reg_field *field, uint32_t val)
+{
+    return (int)((val & field->fld_mask) >> field->fld_pos);
+}
+
+static bool decode_fields;
```

Comment:
Does this need to be global? If so, I recommend making the symbol name less generic.
Eg da1469x_charger_cli_decode_fields. Or even da1469x_cc_decode_fields.
Also I'd recommend moving this global declaration to start of the file.